Test/tutorial/parser_combinators.dfy

```diff
@@ -0,0 +1,294 @@
+/// # Parser combinators in Dafny
+///
+/// Challenge: define parse combinators.  As a usability test, use them to parse
+/// expressions of the form `(ⁿ)ⁿ` (`n` opening parentheses followed by `n`
+/// closing parentheses).
+///
+/// This challenge is tricky because we can't write what we'd usually write in
+/// Haskell or OCaml:
+///
+/// ```ocaml
+/// let parentheses' () =
+///   (either
+///      (concat (char '(') (concat (fun () -> parentheses' ()) (char ')')))
+///      epsilon)
+/// let parentheses () =
+///   (concat parentheses' eos)
+/// ```
+///
+/// Dafny will reject this definition because it cannot prove that the
+/// `parentheses'` function terminates.  This is because Dafny analyses
+/// anonymous functions (“lambdas”) modularly: every time a lambda is created,
+/// as with `(fun () -> parentheses' ())`, Dafny checks that it can be call in
+/// any context.  To see why, consider the function `Apply` below and the
+/// following two uses of it:
+
+function method Apply(f: () -> int): int { f() }
+
+function F0(): int { Apply(F0) }
+function F1(): int { Apply(() => F1()) }
+
+/// Dafny rejects `F0` because it does not allow using “naked” recursive
+/// functions, so one must “eta-expand” (or “thunk”) it, as done in `F1`; but
+/// Dafny also rejects `F1`, because it is not always safe to call (in OCaml, it
+/// would loop forever).  The `parentheses'` parser above *is* always safe to
+/// call, but Dafny doesn't know that, so we need to do a bit more work to
+/// *prove* that it does.
+///
+/// We do it by using stronger specifications to enforce progress, as
+/// shown in the solution below:
+
+module {:options "-functionSyntax:4"} Parsers {
+
+/// (The `-functionSyntax:4` enables Dafny's new function syntax, where
+/// `function` replaces `function method`)
+///
+/// ## Utility types
+///
+/// `Either<L, R>` represents the result of the `Either` combinator.  For `And`,
+/// we just use built-in pairs.
+
+  datatype Either<+L, +R> = Left(l: L) | Right(r: R)
```

Comment:
Is this tutorial going to make use of subtyping? If not, I would suggest avoiding advanced programming concepts which are orthogonal to the tutorial.

Reply:
I prefer to keep the `+` in there, because that's the "right" definition of `Either`.  Is that what you are referring to?